### PR TITLE
added the option for multiple binsizes

### PR DESCRIPTION
**Pull request by yankev**
@cldougl 
Probably need to write a test.

### CODE REVIEW

**Comment by yankev** [comment]
Also have to fix backwards compatibility, because old examples use single floats and not arrays

**Comment by yankev** [comment]
@theengineear Hi Andrew, Is it possible for you to take a look at this PR while Chelsea is away?

**Comment by theengineear** [comment]
Yep. Sure can :). I'll get to it today.

File: plotly/tests/test_optional/test_figure_factory.py [hunk 1]
```diff
@@ -149,6 +149,53 @@ def test_distplot_more_args(self):
                                    'yaxis': 'y1'}
         self.assertEqual(dp['data'][1], expected_dp_data_hist_2)
 
+        def test_distplot_binsize_array(self):
+            hist1_x = [0.8, 1.2, 0.2, 0.6, 1.6,
+                       -0.9, -0.07, 1.95, 0.9, -0.2,
+                       -0.5, 0.3, 0.4, -0.37, 0.6]
+            hist2_x = [0.8, 1.5, 1.5, 0.6, 0.59,
+                       1.0, 0.8, 1.7, 0.5, 0.8,
+                       -0.3, 1.2, 0.56, 0.3, 2.2]
```
Comment (theengineear) [inline plotly/tests/test_optional/test_figure_factory.py]: 🐄 any reason to wrap these lists before the `79` character line length limit?

File: plotly/tests/test_optional/test_figure_factory.py [hunk 2]
```diff
@@ -149,6 +149,53 @@ def test_distplot_more_args(self):
...
+
+            hist_data = [hist1_x] + [hist2_x]
```
Comment (theengineear) [inline plotly/tests/test_optional/test_figure_factory.py]: 🐄 `hist_data = [hist1_x, hist2_x]` would be clearer, IMO.

File: plotly/tests/test_optional/test_figure_factory.py [hunk 3]
```diff
@@ -422,7 +469,7 @@ def test_dendrogram_random_matrix(self):
         self.assert_dict_equal(dendro['layout'], expected_dendro['layout'])
 
     def test_dendrogram_orientation(self):
-        X = np.random.rand(5, 5) 
+        X = np.random.rand(5, 5)
```
Comment (theengineear) [inline plotly/tests/test_optional/test_figure_factory.py]: Your IDE probably did this automatically on save or something, but it's helpful if you *don't* introduce changes to lines of code which are unrelated to the problem you're addressing. Not a huge deal, but something to be mindful of.

File: plotly/tools.py [hunk 4]
```diff
@@ -2332,7 +2332,7 @@ def create_candlestick(open, high, low, close,
 
     @staticmethod
     def create_distplot(hist_data, group_labels,
-                        bin_size=1., curve_type='kde',
+                        bin_size=[1.], curve_type='kde',
```
Comment (theengineear) [inline plotly/tools.py]: This default argument is mutable, which is subject to 🐛s http://docs.python-guide.org/en/latest/writing/gotchas/#mutable-default-arguments

I'd suggest making it `(1.,)` or `None` and checking if it's `None` in the code.

Comment (theengineear) [inline plotly/tools.py]: On second thought, why not just leave it as `1.`? We need to be backwards compatible anyhow.

File: plotly/tools.py [hunk 5]
```diff
@@ -2347,7 +2347,7 @@ def create_distplot(hist_data, group_labels,
         :param (list[list]) hist_data: Use list of lists to plot multiple data
             sets on the same plot.
         :param (list[str]) group_labels: Names for each data set.
-        :param (float) bin_size: Size of histogram bins. Default = 1.
+        :param (list[float]) bin_size: Size of histogram bins. Default = [1.]
```
Comment (theengineear) [inline plotly/tools.py]: Re: https://github.com/plotly/plotly.py/pull/455/files#r63061447

Then the type should be `(list[float]|float)`

File: plotly/tools.py [hunk 6]
```diff
@@ -3931,4 +3934,3 @@ def make_table_annotations(self):
                         font=dict(color=font_color),
                         showarrow=False))
         return annotations
-
```
Comment (theengineear) [inline plotly/tools.py]: Looks like your IDE ate the terminating newline here. Pls put it back :)

File: plotly/tests/test_optional/test_figure_factory.py [hunk 7]
```diff
@@ -149,6 +149,53 @@ def test_distplot_more_args(self):
...
+                                       'opacity': 0.7,
+                                       'type': 'histogram',
+                                       'x': [0.8, 1.5, 1.5, 0.6, 0.59, 1.0, 0.8,
+                                             1.7, 0.5, 0.8, -0.3, 1.2, 0.56, 0.3,
+                                             2.2],
+                                       'xaxis': 'x1',
+                                       'xbins': {'end': 2.2, 'size': 0.2,
+                                                 'start': -0.3},
+                                       'yaxis': 'y1'}
+            self.assertEqual(dp['data'][1], expected_dp_data_hist_2)
```
Comment (theengineear) [inline plotly/tests/test_optional/test_figure_factory.py]: Great test!

**Comment by theengineear** [comment]
Awesome! Couple of comments for ya! Let me know if you have any questions and ping me when you're ready for another review.

File: plotly/tests/test_optional/test_figure_factory.py (hunk 1 above)
Comment (yankev) [inline plotly/tests/test_optional/test_figure_factory.py]: Oh no reason, I just copy&pasted that data from another distplot test. Though I can definitely save a line by following that tip.

File: plotly/tests/test_optional/test_figure_factory.py (hunk 2 above)
Comment (yankev) [inline plotly/tests/test_optional/test_figure_factory.py]: k, I agree.

File: plotly/tests/test_optional/test_figure_factory.py (hunk 3 above)
Comment (yankev) [inline plotly/tests/test_optional/test_figure_factory.py]: Right, I have no idea why that happened. Will do.

File: plotly/tools.py (hunk 6 above)
Comment (yankev) [inline plotly/tools.py]: k, ye I just started using `atom`, and it doesn't seem to raise an issue about that anymore.

**Comment by yankev** [comment]
@theengineear Should be good to go!

File: plotly/tools.py (hunk 6 above)
Comment (theengineear) [inline plotly/tools.py]: Please just add a blank newline back to the end of this file still.

**Comment by theengineear** [comment]
@yankev great! Nice work! Just address https://github.com/plotly/plotly.py/pull/455#discussion_r63133901 and after you've done that 💃 ! Merge away.

**Comment by yankev** [comment]
@theengineear k added! Thanks for the help!

**Pull request merged**